### ISSUE TITLE
Log flooded with "shutting down kafka sink" after a Kafka problem

### ISSUE THREAD

**Issue by philip-stoev** [MaterializeInc/materialize#8561]
### What version of Materialize are you using?

```
v0.9.6-dev (8c4d31411)
```

### What was the issue?

Feeding large amounts of data into 5 sinks (all with reuse_topic=true) in case that matters, at some point the log said:

```
2021-10-06T12:32:29.891995Z ERROR dataflow::sink::kafka: encountered error during kafka interaction. kafka-u17 in state BeginTxn at time 1633523534606 : Transaction error: Operation not valid in state Begin
Commit    
```

followed by an incessant stream of 10-per-second messages like this:

```
2021-10-06T12:32:31.801875Z  INFO dataflow::sink::kafka: shutting down kafka sink: kafka-u17    
2021-10-06T12:32:32.032769Z  INFO dataflow::sink::kafka: shutting down kafka sink: kafka-u17    
2021-10-06T12:32:32.113915Z  INFO dataflow::sink::kafka: shutting down kafka sink: kafka-u17    
2021-10-06T12:32:32.128084Z  INFO dataflow::sink::kafka: shutting down kafka sink: kafka-u17    
2021-10-06T12:32:32.316834Z  INFO dataflow::sink::kafka: shutting down kafka sink: kafka-u17    
2021-10-06T12:32:32.380275Z  INFO dataflow::sink::kafka: shutting down kafka sink: kafka-u17    
2021-10-06T12:32:32.426443Z  INFO dataflow::sink::kafka: shutting down kafka sink: kafka-u17    
2021-10-06T12:32:32.629094Z  INFO dataflow::sink::kafka: shutting down kafka sink: kafka-u17    
2021-10-06T12:32:32.693051Z  INFO dataflow::sink::kafka: shutting down kafka sink: kafka-u17    
2021-10-06T12:32:32.732947Z  INFO dataflow::sink::kafka: shutting down kafka sink: kafka-u17    
2021-10-06T12:32:32.937700Z  INFO dataflow::sink::kafka: shutting down kafka sink: kafka-u17    
2021-10-06T12:32:32.941405Z  INFO dataflow::sink::kafka: shutting down kafka sink: kafka-u17    
2021-10-06T12:32:32.962608Z  INFO dataflow::sink::kafka: shutting down kafka sink: kafka-u17    
2021-10-06T12:32:33.033397Z  INFO dataflow::sink::kafka: shutting down kafka sink: kafka-u17    
```

### Is the issue reproducible? If so, please provide reproduction instructions.

I can provide a test case but I hope this is solveable by code inspection alone.

**Comment by ruchirK** [comment]
@philip-stoev I don't think that commit is present on main? I'm pretty sure that commit is from an old version of https://github.com/MaterializeInc/materialize/pull/8434 before the last, crucial edits that Aljoscha requested. Can you check if this issue still exists on latest main -- if so, I'll happily debug ASAP

**Comment by nmeagan11** [comment]
@philip-stoev does this issue still exist?

**Comment by cjubb39** [comment]
This should be fixed by #14965.  We currently hit the branch with this print in two situations:
1. The `KafkaSinkToken` has been dropped indicating we've gotten a `DROP SINK` command -- or we're genuinely shutting down.  In that case, we don't have any expectation of hitting this print more than a couple times (but likely once -- depending on how the operator is scheduled I suppose)
2. The sink hits an error and moved into a state where it doesn't do anything but still has a running dataflow.  It may still be invoked again after this when the input frontier changes so we repeatedly print the message.  This case will be addressed by the PR I mentioned (which, incidentally doesn't print anything other than the panic message)
